src/components/Topology/Topology.js

```diff
@@ -240,25 +320,158 @@ export default ({ modelData, width, height }) => {
+      //       type: "success",
+      //       message: "The position of your apps have been updated.",
+      //     });
+      //   }
+      // });
+    }
+
+    if (!isReadOnly) {
+      appIcon.call(
+        d3.drag().on("start", dragstarted).on("drag", drag).on("end", dragended)
+      );
+    }
+
+    setRelations(relationLine);
 
     appIcons.exit().remove();
     relationLines.exit().remove();
 
     return () => {
       topo.remove();
     };
-  }, [applications, deltaX, deltaY, height, width, maxX, maxY, relations]);
-  return <svg ref={ref} />;
+  }, [
+    applications,
+    conn,
+    deltaX,
+    deltaY,
+    isFullscreen,
+    isReadOnly,
+    maxX,
+    maxY,
+    relations,
+  ]);
+
+  // XXX Remove flash messaging until we can save results back to juju
+  // Reset flash messages after 3 seconds
+  // useEffect(() => {
+  //   if (flashMessage.message) {
+  //     window.clearFlashMessage = setTimeout(() => {
+  //       setFlashMessage({});
+  //     }, 3000);
+  //   }
+  //   return () => {
+  //     clearTimeout(window.clearFlashMessage);
+  //   };
+  // }, [flashMessage]);
+
+  // Close topology, if open, on Escape key press
+  useEffect(() => {
+    const closeOnEscape = function (e) {
+      if (e.code === "Escape" && isFullscreen) {
+        setIsFullscreen(false);
+      }
+    };
+    document.addEventListener("keydown", closeOnEscape);
+    return () => {
+      document.removeEventListener("keydown", closeOnEscape);
+    };
+  });
+
+  console.log("texts");
+  if (Object.entries(modelData?.applications || {}).length) {
+    return (
+      <div
+        className="topology"
+        data-fullscreen={isFullscreen}
+        data-readonly={isReadOnly}
+        ref={topologyRef}
+      >
+        <div className="topology__inner">
+          {/* Remove flash messaging until we can save positions back to juju */}
+          {/* <div className="flashMessage" data-type={flashMessage?.type}>
+            {flashMessage?.message}
+          </div> */}
+          {isFullscreen && (
+            <i
+              className="p-icon--close"
+              onClick={() => setIsFullscreen(false)}
+            />
+          )}
+
+          {modelData ? (
+            <div className="topology__svg">
+              <svg ref={svgRef} />
+            </div>
+          ) : (
+            <div className="topology__loading">
+              <i className="p-icon--spinner">Loading...</i>
+            </div>
+          )}
+
+          {!isFullscreen && modelData && (
+            <i
+              style={{ backgroundImage: `url(${fullScreenIcon})` }}
+              className="p-icon--expand p-icon--fullscreen"
+              data-test="icon--fullscreen"
+              onClick={() => setIsFullscreen(true)}
+            />
+          )}
+          {isReadOnly && isFullscreen && (
+            <span className="read-only">Read only</span>
+          )}
+        </div>
+      </div>
+    );
+  } else {
+    return (
```

Comment:
This doesn't look great, Could we have just an empty grey box rendered while it's waiting for data?

Reply:
Then what do we do for models without any applications?